core/src/apps/zcash/sig_hasher.py

```diff
@@ -0,0 +1,187 @@
+from trezor.crypto.hashlib import blake2b
+from trezor.utils import HashWriter, empty_bytearray
+from apps.bitcoin.writers import (
+    write_bytes_fixed,
+    write_bytes_reversed,
+    write_tx_output,
+    write_uint32,
+    write_uint64,
+    write_bytes_prefixed,
+    write_uint8,
+    TX_HASH_SIZE,
+)
+from apps.bitcoin import scripts
+from apps.bitcoin.common import SigHashType
+
+from typing import TYPE_CHECKING
+if TYPE_CHECKING:
+    from trezor.messages import TxInput, TxOutput, SignTx, PrevTx
+    from trezor.utils import Writer
+    from apps.common.coininfo import CoinInfo
+    from typing import Sequence
+
+
+def write_hash(w: Writer, hash: bytes):
+    write_bytes_fixed(w, hash, TX_HASH_SIZE)
+
+
+def write_prevout(w: Writer, txi: TxInput):
+    write_bytes_reversed(w, txi.prev_hash, TX_HASH_SIZE)
+    write_uint32(w, txi.prev_index)
+
+
+class ZcashSigHasher:
+    def __init__(self):
+        self.header = HeaderHasher()
+        self.transparent = TransparentHasher()
+        self.sapling = SaplingHasher()
+        self.orchard = OrchardHasher()
+
+        self.tx_hash_person = None
+        self.initialized = False
+
+    def initialize(self, tx: SignTx | PrevTx):
```

Comment:
I am not too fond of these `initialize` methods. I think the proper way is for this to go in `__init__()`. I assume the reason why it's not there is because `Signer.create_sig_hasher()` doesn't have the `tx` parameter, so one solution would be:
- Add the `tx` parameter to `create_sig_hasher()` and `__init__()` and to remove it from `hash341()` and `hash143()`.

However, the sighasher doesn't actually need to be initialized with the `tx` parameter. It is only used to precompute the header digest and some personalization string, which can both be computed in `signature_digest()` and `txid_digest()` and optionally cached to avoid recomputing. So another option is:
- Add the `tx` parameter to `signature_digest()` and `txid_digest()` and cache the header digest.

Reply:
I don't like an idea of passing `tx` every time some digest is needed while using it only once. I will pass it as argument of `create_sig_hasher()`.

Reply:
https://github.com/trezor/trezor-firmware/pull/2166/commits/976d716b73f53089fda19981d27c8be3041c391c and https://github.com/trezor/trezor-firmware/pull/2166/commits/4abffe5215e1db8d972ae1ae29b81d8be4323c5f